cmd/csi-api-gen/generators/api.go

```diff
@@ -0,0 +1,432 @@
+	//		root package
+	// * +csi-proxy-api-gen=groupName:<snake_case_group_name> to set the group's name
+	//   	- defaults to the package's name
+	// * +csi-proxy-api-gen=serverBasePkg:<pkg_path> to set the base output directory
+	//		for generated server files - defaults to github.com/kubernetes-csi/csi-proxy/internal/server
+	// * +csi-proxy-api-gen=clientBasePkg:<pkg_path> to set the base output directory
+	//		for generated client files - defaults to github.com/kubernetes-csi/csi-proxy/client/groups
+	tagMarker = "+"
+	tagName   = "csi-proxy-api-gen"
+
+	// headerComment is the comment on the first line of every generated file.
+	headerComment = "// Code generated by csi-proxy-api-gen. DO NOT EDIT.\n\n"
+
+	// pkgPlaceholder can be used as a placeholder for the package path, e.g.
+	// when deriving what callbacks an API group's internal server must have
+	// in order to be able to service all its versions.
+	// See also replaceTypesPackage.
+	pkgPlaceholder = "__PKG_PLACEHOLDER__"
+)
+
+// nameSystems returns the name system used by the generators in this package.
+func nameSystems() namer.NameSystems {
+	return namer.NameSystems{
+		"public":        namer.NewPublicNamer(0),
+		"removePackage": &removePackageNamer{},
+		"short":         &shortNamer{},
+	}
+}
+
+// defaultNameSystem returns the default name system for ordering the types to be
+// processed by the generators in this package.
+func defaultNameSystem() string {
+	return "public"
+}
+
+// packages returns a list of package generators.
+func packages(context *generator.Context, arguments *args.GeneratorArgs) (pkgs generator.Packages) {
+	// find API group definitions
+	groups := findAPIGroupDefinitions(context)
+	klog.V(5).Infof("Found API groups: %v", groups)
+
+	for _, group := range groups {
+		removeGeneratedFiles(group, arguments.OutputBase)
+		pkgs = append(pkgs, packagesForGroup(group, arguments.OutputBase)...)
+	}
+	return
+}
+
+// findAPIGroupDefinitions iterates over the context's list of package paths,
+// and builds a map mapping API group paths to their definition.
+// API group definitions are either:
+// * subdirectories of client/api
+// * or packages whose doc.go file contains a comment containing relevant tag markers
+// If the latter, see the comments on tagMarker and tagName above for supported options.
+func findAPIGroupDefinitions(context *generator.Context) []*groupDefinition {
+	pkgPaths := context.Inputs
+
+	// first, re-order the inputs by lengths, so that we always process parent packages first
+	sort.Slice(pkgPaths, func(i, j int) bool {
+		return len(pkgPaths[i]) < len(pkgPaths[j])
+	})
+
+	groups := make(map[string]*groupDefinition)
+
+	for _, pkgPath := range pkgPaths {
+		klog.V(5).Infof("Considering input %s", pkgPath)
+
+		pkg := context.Universe[pkgPath]
+		if pkg == nil {
+			// If the input had no Go files, for example.
+			continue
+		}
+
+		if buildAPIGroupDefinitionFromDocComment(pkgPath, pkg, groups) {
```

Comment:
so it is not possible to have group definition in both subdirectories of client/api and doc.go file ?

Reply:
It's absolutely possible. If a group definition is in the canonical client/api/<name> location _and_ has a doc.go file with `csi-proxy-api-gen` tags, then it's processed according to the tags' options (and if there are no custom options then the tag is redundant with it being in the canonical location)

Reply:
I am trying to check the continue logic here means if buildAPIGroupDefinitionFromDocComment returns true, the rest will be skipped?

Reply:
Yes, it will, since the `groupDefinition` for it will already have been built.